components/post_view/reaction/reaction.jsx

```diff
@@ -10,88 +10,152 @@ import OverlayTrigger from 'components/overlay_trigger';
+        if (nextProps.reactionCount !== this.props.reactionCount) {
+            // if current user is reacting
+            if (!this.props.sortedUsers.currentUserReacted && nextProps.sortedUsers.currentUserReacted) {
+                this.setState({
+                    displayNumber: nextProps.reactionCount,
+                    reactedNumber: nextProps.reactionCount,
+                    unreactedNumber: nextProps.reactionCount - 1,
+                });
+                return true;
+            }
+
+            // if current user is unreacting
+            if (this.props.sortedUsers.currentUserReacted && !nextProps.sortedUsers.currentUserReacted) {
+                this.setState({
+                    displayNumber: nextProps.reactionCount,
+                    reactedNumber: nextProps.reactionCount + 1,
+                    unreactedNumber: nextProps.reactionCount,
+                });
+                return true;
+            }
+
+            // if another user is reacting/unreacting
+            if (this.props.sortedUsers.currentUserReacted === nextProps.sortedUsers.currentUserReacted) {
+                if (this.props.sortedUsers.currentUserReacted) {
+                    this.setState({
+                        displayNumber: nextProps.reactionCount,
+                        reactedNumber: nextProps.reactionCount,
+                        unreactedNumber: nextProps.reactionCount - 1,
+                    });
+                    return true;
+                }
+                this.setState({
+                    displayNumber: nextProps.reactionCount,
+                    reactedNumber: nextProps.reactionCount + 1,
+                    unreactedNumber: nextProps.reactionCount,
+                });
+                return true;
+            }
+        }
+        return true;
+    }
+
+    handleClick = () => {
+        // only proceed if user has permission to react
+        if (!this.props.canAddReaction || !this.props.canRemoveReaction) {
+            return;
+        }
+        this.setState((state) => {
+            if (state.userReacted) {
+                return {
+                    userReacted: false,
+                    reactedClass: 'Reaction unreacting',
+                };
+            }
+            return {
+                userReacted: true,
+                reactedClass: 'Reaction reacting',
+            };
+        });
+    }
+
+    handleAnimationEnded = () => {
+        const {reactedNumber, unreactedNumber} = this.state;
+        const {actions, post, emojiName} = this.props;
+        this.setState((state) => {
+            if (state.userReacted) {
+                return {
+                    reactedClass: 'Reaction reacted',
+                    displayNumber: reactedNumber,
+                };
+            }
+            return {
+                reactedClass: 'Reaction unreacted',
+                displayNumber: unreactedNumber,
+            };
+        });
+        if (this.state.userReacted) {
```

Comment:
Perhaps it would be better to do this in response to the `setState` call in `handleClick` using `setState`'s callback parameter? I didn't expect to see these actions being fired in response to something other than the click event.

I suppose that could cause issues if the callback returns while we're still animating and updates the props though.

Reply:
The primary reason for calling here is to prevent the app from repainting during the animation. In Chrome Dev Tools calling the redux action during the animation caused re-painting in several places and slow the animation down to just a few frames.

Reply:
What about using the `setState` callback to trigger it then? Does that still cause the repaint?

Reply:
I can try it out locally, but my guess is anything that start to change redux state during the animation is going to slow down FPS because of the non-UI react lifecycle processing plus the UI repaint. Is there a reason you're concerned about waiting until the animation is done?

Reply:
I inspected the performance and I'm convinced the animation should fire first, with the action triggered immediately after. I can see a ton of recomputes in the virtual DOM that do surge the CPU/GPU performance. The result is a drop in frame-rate. It is subtle, but the animation is so short it does make a difference (GIF at 30fps):
![anim-comparison](https://user-images.githubusercontent.com/44858354/83086071-96697200-a042-11ea-8398-f29f00d31245.gif)